src/Tools/dotnet-monitor/Profiler/ProfilerService.cs

```diff
@@ -0,0 +1,176 @@
+            {
+                // This condition occurs when runing dotnet-monitor from the build output because
+                // the shared libraries are colocated into the "shared" folder due to packaging, not
+                // at build time. This setting is valid for testing scenarios.
+                // CONSIDER: Remove test-specific code and conditions, if possible, and replace with
+                // some other extensible mechanism.
+                _libraryPathSource.SetResult(null);
+            }
+        }
+
+        public async Task ApplyProfiler(IEndpointInfo endpointInfo, CancellationToken cancellationToken)
+        {
+            try
+            {
+                DiagnosticsClient client = new DiagnosticsClient(endpointInfo.Endpoint);
+                Dictionary<string, string> env = await client.GetProcessEnvironmentAsync(cancellationToken);
+
+                // CONSIDER: Allow alternate means of specifying or heuristically determining the runtime identifier
+                // - For Windows and OSX, build identifier from platform + architecture of target process.
+                // - Lookup same environment variable on dotnet-monitor itself.
+                // - Check libc type of dotnet-monitor.
+                env.TryGetValue(DotnetMonitorProfiler_RuntimeIdentifier, out string runtimeIdentifier);
+
+                if (string.IsNullOrEmpty(runtimeIdentifier))
+                {
+                    throw new InvalidOperationException("Unable to determine platform of the target platform.");
+                }
+                else
+                {
+                    // This may be null if running dotnet-monitor from build output instead of from
+                    // dotnet tool installation.
+                    string sharedLibraryPath = await _libraryPathSource.Task;
+
+                    IFileProvider nativeFileProvider;
+                    if (string.IsNullOrEmpty(sharedLibraryPath))
+                    {
+                        // CONSIDER: Allow some way to of specifying the build output location of the libraries
+                        // without have to code it into the product.
+                        nativeFileProvider = NativeFileProvider.CreateTest(runtimeIdentifier);
+                    }
+                    else
+                    {
+                        nativeFileProvider = NativeFileProvider.CreateShared(runtimeIdentifier, sharedLibraryPath);
+                    }
+
+                    string libraryName = NativeLibraryHelper.GetSharedLibraryName("MonitorProfiler");
+
+                    IFileInfo profilerFileInfo = nativeFileProvider.GetFileInfo(libraryName);
+                    if (!profilerFileInfo.Exists)
+                    {
+                        // Do not use IFileInfo.PhysicalPath as that is null of files that do not exist.
+                        throw new FileNotFoundException("Could not find profiler assembly at determined path.", profilerFileInfo.Name);
```

Comment:
Should this be `profilerFileInfo.FullName`, or do we specifically only want the `Name`?

Reply:
[IFileInfo](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.fileproviders.ifileinfo?view=dotnet-plat-ext-6.0) does not have a `FullName` property. Note that this is an interface named `IFileInfo` from the `Microsoft.Extensions.FileProviders.Abstractions` assembly, not the class named `FileInfo` from the `System.Runtime` assembly.

Reply:
Ah, that was an oversight on my part, sorry about that.